nexus/src/db/datastore/project.rs

```diff
@@ -65,33 +66,241 @@ impl DataStore {
         })
     }
 
+    async fn ensure_no_instances_in_project(
+        &self,
+        opctx: &OpContext,
+        authz_project: &authz::Project,
+    ) -> DeleteResult {
+        use db::schema::instance;
+
+        // Make sure there are no instances present within this project.
+        let instance_found = diesel_pool_result_optional(
+            instance::dsl::instance
+                .filter(instance::dsl::project_id.eq(authz_project.id()))
+                .filter(instance::dsl::time_deleted.is_null())
+                .select(instance::dsl::id)
+                .limit(1)
+                .first_async::<Uuid>(self.pool_authorized(opctx).await?)
+                .await,
+        )
+        .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))?;
+        if instance_found.is_some() {
+            return Err(Error::InvalidRequest {
+                message: "project to be deleted contains an instance"
```

Comment:
Would it be possible (or, easy) to include information about the thing that we found in the internal message?  I'm thinking about debugging a code path like this (e.g., we find a case where it's unexpectedly failing to delete a Project because some Instance is still around, but we don't think there's an Instance around).  It could make a world of difference if we knew which instance Nexus thought was keeping this Project alive, particularly if the problem winds up being transient.

Same with the other functions.

Reply:
Done in 31ec7651c429503deb11d06b966da9eb5592fc11